### PR TITLE
feat: Swap Fees Paid Contribute to Tx Fee for Sybil Resistant Fees

### PR DESCRIPTION
**Pull request by lostak**
Closes: #1139   
[Previous PR](https://github.com/osmosis-labs/osmosis/pull/1675)

## What is the purpose of the change

Swap fees paid for a msg are added to Tx fee when determining if the fee is sufficient for the gas price
- If the tx fee attached for a swap message is not sufficient to pay for gas, the swap fees paid can make up the difference
- The attached tx fee is always deducted in its entirety if it exists
- If the gas price is zero, no tx fee needs to be attached
- If the message satisfies the SybilResistantFee interface, add the swap fees paid to the sybil struct
- For all tx, add tx fee to sybil fees paid (for non-applicable msgs, feed paid = 0)
- Compare sybil fees paid to gas price when determining if a fee is sufficient

## Brief Changelog

- Add `GetSwapFee(ctx sdk.Context, poolId uint64) (sdk.Dec, error)` to `swap.go` and to the `GammKeeper` in the `expected_keepers` for the `txfees` module
- Add interface `SybilResistantFee` in a new file `msg_sybil.go` in `gamm/types/` 
- Add implement `SybilResistantFee` in all msgs involving a swap, including multihops
- Add file `sybil.go` for sybil resistant fees struct and methods
- `AnteHandle` now treats `SybilResistantFee` msgs differently when checking a fee but other msgs use the same logic as before

## Testing and Verifying

`make test`, `make test-sim`, `make test-unit` and `make build` are all successful and specific tests are being written

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not documented)

### CODE REVIEW

**Comment by lostak** [comment]
[Diagram Comparing Proposed Changes in Logic and Files to the Current Fee Decorator Process](https://github.com/osmosis-labs/osmosis/files/8934664/SybilResistantFeeDiagram.pdf)

File: x/gamm/types/msg_sybil.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+// SybilResistantFee Swap Msg defines an interface for getting
+// - poolIds on a swap msg route
+// - token to apply the sybil resistant fees to
+type SybilResistantFee interface {
+	GetTokenDenomsOnPath() []string
+	GetPoolIdOnPath() []uint64
+	GetTokenToFee() sdk.Coin
+}
```
Comment (ValarDragon) [inline x/gamm/types/msg_sybil.go:14]: I think we should just add the `GetPoolIdOnPath` to the existing `msg_swap.go` interface. Not sure why "GetTokenToFee" exists

Comment (lostak) [reply x/gamm/types/msg_sybil.go:14]: By using a separate interface, `SybilResistantFee` can be more abstractly applied to any message involving a swap, whereas `SwapMsgRoute` is only satisfied by messages with swap routes in their proto file.
[These lines](https://github.com/osmosis-labs/osmosis/blob/d7fa2fd172668b4a0eb7d8207b3e2d0c952d43b6/x/txfees/keeper/feedecorator.go#L167-L185) of `feedecorator.go` use this abstraction and could work for msgs like `MsgJoinSwapExternAmountIn` properly satisfying `SybilResistantFee`. 
`GetTokenToFee()` returns the `sdk.Coin` that would have paid the swap fees, in its initial amount. This is then used to check if the token is applicable for use as a `feeToken`, and to compute the fees paid in the base denom.
If the swap is a `multihop`, then only swap fees from pools with a `feeToken` are added to the total swap fee.

Comment (lostak) [reply x/gamm/types/msg_sybil.go:14]: If you only see this as being used for swaps, not joins and exits, I'll move the new functionality to the existing interface.

**Pull request closed without merging**